activesupport/lib/active_support/message_verifier.rb

```diff
@@ -106,18 +118,36 @@ def verify(signed_message)
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
     #   verifier.generate 'a private message' # => "BAhJIhRwcml2YXRlLW1lc3NhZ2UGOgZFVA==--e2d724331ebdee96a10fb99b089508d1c72bd772"
-    def generate(value)
-      data = encode(@serializer.dump(value))
-      "#{data}--#{generate_digest(data)}"
+
+    def generate(options)
+      @claims = Claims.new(options)
+      data = encode(encoded_header, encoded_claims).join('.')
+      "#{data}.#{generate_digest(data)}"
     end
 
     private
-      def encode(data)
-        ::Base64.strict_encode64(data)
+      def header
+        { 'typ' => 'JWT', 'alg' => @digest.to_s }
+      end
+
+      def encoded_header
+        @serializer.dump header
+      end
+
+      def encoded_claims
+        @serializer.dump @claims.to_h
+      end
+
+      def encode(*args)
```

Comment:
I don't see the need for the changes in `encode` and `decode` - at the moment the change is more a pain than worth it.